### PR TITLE
Migrate EC support to Rust

### PR DESCRIPTION
**Pull request by alex**

### CODE REVIEW

**Comment by alex** [comment]
Note to self: Do we have a test that verifies that we always serialize in named curve form?

File: src/rust/src/backend/ec.rs [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+                exceptions::UnsupportedAlgorithm::new_err((
+                    "Unsupported EC exchange algorithm",
+                    exceptions::Reasons::UNSUPPORTED_EXCHANGE_ALGORITHM,
+                )),
+            ));
+        }
+
+        let mut deriver = openssl::derive::Deriver::new(&self.pkey)?;
+        #[cfg(CRYPTOGRAPHY_OPENSSL_300_OR_GREATER)]
+        deriver
```
Comment (reaperhulk) [inline src/rust/src/backend/ec.rs:332]: Worth a comment here to explain why we don't re-validate?

File: src/rust/src/backend/ec.rs [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+                ),
+            )?
+            .extract()?;
+
+        let mut verifier = openssl::pkey_ctx::PkeyCtx::new(&self.pkey)?;
+        verifier.verify_init()?;
+        let valid = verifier.verify(data, signature).unwrap_or(false);
+        // TODO: Empty the error stack. BoringSSL leaves one in the event of
+        // signature validation failure. Upstream to rust-openssl?
+        openssl::error::ErrorStack::get();
```
Comment (reaperhulk) [inline src/rust/src/backend/ec.rs:502]: Should we annotate this for just boring?

Comment (alex) [reply src/rust/src/backend/ec.rs:506]: Done, but also, what do you think about upstreaming?

Comment (reaperhulk) [reply src/rust/src/backend/ec.rs:506]: I think we should upstream. Without doing so every caller needs to know that they need this conditional or else their error stack gets polluted and who knows what their error handling does.

**Review by reaperhulk — APPROVED** [review]

**Pull request merged**